### ISSUE TITLE
Issue with using read_image for reading segmentation masks

### ISSUE THREAD

**Issue by ashnair1** [facebookresearch/detectron2#1760]
Since commit ffe4ca12727d715143fd387f0c7ce25a991d8e95 I keep running into errors while reading my segmentation masks for semantic segmentation. I've isolated the problem to this line [here](https://github.com/facebookresearch/detectron2/blob/e9be1d7190122a691a82cd31ec3354b7cd4203d2/detectron2/data/detection_utils.py#L75). In this case, the format is 'L' and the image gets converted to that format before being converted to an array. However, this causes my segmentation masks to be zeroed out. I tried using the previous method of reading segmentation masks and it works fine. 

So I'm unsure how to proceed. It could be due to the way I've stored the segmentation masks. But I was able to train and test a segmentation model prior to this commit and it worked. Any recommendations?

**Comment by ppwwyyxx** [comment]
Could you provide an example of your segmentation mask?

**Comment by ashnair1** [comment]
I see. My masks are in palette mode. Here are some examples:

![P1360_600_1400_3200_4000](https://user-images.githubusercontent.com/32295036/87223038-928f7600-c38a-11ea-8279-41bc312ed461.png)

![P1164_1200_2000_0_800](https://user-images.githubusercontent.com/32295036/87223020-4a705380-c38a-11ea-994e-9775e7fdd65b.png)

**Comment by ppwwyyxx** [comment]
That is not a supported format. It should be a grayscale image where each pixel is an integer (therefore no color). See https://detectron2.readthedocs.io/tutorials/datasets.html

**Comment by ashnair1** [comment]
@ppwwyyxx Just a follow up question, is detectron2 open to supporting palette mode? I can only speak for myself but having the segmentation masks (where pixel values are integer labels) with the palette makes visualising the ground truth easier as opposed to the current grayscale standard where all the segmentation masks will be dark (assuming your num_classes < 255).

**Comment by ppwwyyxx** [comment]
You can use colored images in the ground truth, as long as you also [implement data loader](https://detectron2.readthedocs.io/tutorials/data_loading.html) to convert them to integers, instead of using the default dataloader that will run into the issue you encountered before. This is because every model requires semantic segmentation ground truth to be integers. 

Alternatively, you can use integers in the groundtruth but implement better visualization. For example, detectron2's builtin visualizer will add colors.

**Comment by ashnair1** [comment]
I see. The reason why I'm partial to palette mode is that the modification is at the data level. Using palettes will allow semantic segmentation ground truth to be integers while allowing users to visualise the image with no additional changes to their code. This helps when you're preparing data for training in detectron2.



Proposal: 
Change 2 lines

From
```
if format is not None:
        # PIL only supports RGB, so convert to RGB and flip channels over below
        conversion_format = format
        if format in ["BGR", "YUV-BT.601"]:
            conversion_format = "RGB"
        image = image.convert(conversion_format)
    image = np.asarray(image)
    # PIL squeezes out the channel dimension for "L", so make it HWC
    if format == "L":
        image = np.expand_dims(image, -1)
```


To
```
if format is not None:
        # PIL only supports RGB, so convert to RGB and flip channels over below
        conversion_format = format
        if format in ["BGR", "YUV-BT.601"]:
            conversion_format = "RGB"
        if format != "P":                                  # Don't convert if image is in palette mode
            image = image.convert(conversion_format)
    image = np.asarray(image)
    # PIL squeezes out the channel dimension for "L", so make it HWC
    if format in ["L", "P"]:                            # Add extra dim just like for 'L' images
        image = np.expand_dims(image, -1)
```

**Comment by ppwwyyxx** [comment]
Maybe that can work for now. However palette mode is not common today and many libaries probably don't handle them the same way Pillow does. As a result we probably won't invest in it. 
For example we plan to switch to opencv at some point (https://github.com/facebookresearch/detectron2/issues/698#issuecomment-574871727) and it may affect palette images.

If you have special needs like this it's better to use a custom dataloader.

**Comment by ashnair1** [comment]
Oh I see. Thanks for the tip. I'll go with custom dataloaders then
